benchmarks/benchmark.py

```diff
@@ -0,0 +1,495 @@
+                        var_conversions += "    auto {} = stan::math::to_matrix_cl({});\n".format(var_name + "_cl",
+                                                                                                  var_name)
+                        var_name += "_cl"
+                if opencl == "copy" and stan_arg not in ("int", "real"):
+                    code += "stan::math::to_matrix_cl({}), ".format(var_name)
+                else:
+                    code += var_name + ", "
+            if opencl == "base":
+                var_conversions += "  stan::math::opencl_context.queue().finish();\n"
+            code = code[:-2] + ");\n"
+            if "Rev" in arg_overloads:
+                code += "    stan::test::recursive_sum(res).grad();\n"
+            result += benchmark_template.format(benchmark_name=benchmark_name, setup=setup,
+                                                var_conversions=var_conversions, code=code, multi=multiplier,
+                                                max_size=max_size)
+    cpp_filepath = working_folder + cpp_filename
+    with open(cpp_filepath, "w") as o:
+        o.write("#include <benchmark/benchmark.h>\n")
+        o.write("#include <test/expressions/expression_test_helpers.hpp>\n\n")
+        o.write(result)
+        o.write("BENCHMARK_MAIN();")
+    exe_filepath = cpp_filepath.replace(".cpp", exe_extension)
+    build(exe_filepath)
+    run_benchmark(exe_filepath, n_repeats, csv_out_file)
+
+
+def main(functions_or_sigs, cpp_filename="benchmark.cpp", overloads=("Prim", "Rev"), multiplier_param=None,
+         max_size_param=None, max_dim=3, n_repeats=1, skip_similar_signatures=False, csv_out_file=None, opencl=False,
+         plot=False, plot_log_y=False,
+         plot_cl_speedup=False, plot_reference=None):
+    """
+    Generates benchmark code, compiles it and runs the benchmark. Optionally plots the results.
+    :param functions_or_sigs: List of function names and/or signatures to benchmark
+    :param cpp_filename: filename of cpp file to use
+    :param overloads: Which overloads to benchmark
+    :param multiplier_param: Multiplyer, by which to increase argument size.
+    :param max_size_param: Maximum argument size.
+    :param max_dim: Maximum number of argument dimensions to benchmark. Signatures with any argument with
+             larger number of dimensions are skipped."
+    :param n_repeats: Number of times to repeat each benchmark.
+    :param skip_similar_signatures: Whether to skip similar signatures. Two signatures are similar if they
+             difffer only in similar vector types, which are vector, row_vector and real[].
+    :param csv_out_file: Filename of the csv file to store benchmark results in.
+    :param plot: Filename of bmp or csv fle to store plot into. If  filename is empty, opens a window with graph.
+    :param plot_log_y: Use logarithmic y axis for plotting
+    :param plot_cl_speedup: plot speedup of OpenCL overloads compared to CPU ones
+    """
```

Comment:
```suggestion
    """
    if overloads is None:
        overloads = ("Prim", "Rev")
```

Reply:
Why do you prefer it this way? So far I used this only for defaults I want to depend on other values.

Reply:
It is just a common way to use defaults (sure one can not change values for tuples).